framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,840 @@
+    # otherwise, continue submitting and collecting
+
+  ###################
+  # Utility Methods #
+  ###################
+  def _checkStepReady(self, traj):
+    """
+      Checks if enough information has been collected to proceed with optimization
+      @ In, traj, int, identifier for trajectory of interest
+      @ Out, ready, bool, True if all required data has been collected
+    """
+    # ready to move step forward if (a) we have the opt point, and (b) we have the grad points
+    tracker = self._stepTracker[traj]
+    # check (a) we have the opt point
+    if tracker['opt'] is None:
+      return False
+    # check (b) we have the grad points
+    if len(tracker['grads']) < self._gradientInstance.numGradPoints():
+      return False
+    # if all checks passed, we're ready
+    return True
+
+  def _initializeStep(self, traj):
+    """
+      Initializes a new step in the optimization process.
+      @ In, traj, int, the trajectory of interest
+      @ Out, None
+    """
+    RavenSampled._initializeStep(self, traj)
+    self.incrementIteration(traj)
+    # tracker 'opt' set up in Sampled
+    self._stepTracker[traj]['grads'] = []
+
+  def initializeTrajectory(self, traj=None):
+    """
+      Handles the generation of a trajectory.
+      @ In, traj, int, optional, label to use
+      @ Out, traj, int, new trajectory number
+    """
+    traj = RavenSampled.initializeTrajectory(self)
+    self._gradHistory[traj] = deque(maxlen=self._maxHistLen)
+    self._stepHistory[traj] = deque(maxlen=self._maxHistLen)
+    self._acceptHistory[traj] = deque(maxlen=self._maxHistLen)
+    self._stepRecommendations[traj] = None
+    self._acceptRerun[traj] = False
+    self._convergenceInfo[traj] = {'persistence': 0}
+    for criteria in self._convergenceCriteria:
+      self._convergenceInfo[traj][criteria] = False
+    return traj
+
+  def _resolveNewGradPoint(self, traj, rlz, optVal, info):
+    """
+      Consider and store a new gradient evaluation point
+      @ In, traj, int, trajectory for this new point
+      @ In, info, dict, identifying information about the realization
+      @ In, rlz, dict, realized realization
+      @ In, optVal, float, value of objective variable (corrected for min/max)
+    """
+    self._stepTracker[traj]['grads'].append((rlz, info))
+
+  # * * * * * * * * * * * * * * * *
+  # Resolving potential opt points
+  def _applyBoundaryConstraints(self, point):
+    """
+      Checks and fixes boundary constraints of variables in "point" -> DENORMED point expected!
+      @ In, point, dict, potential point against which to check
+      @ Out, point, dict, adjusted variables
+      @ Out, modded, bool, whether point was modified or not
+    """
+    # TODO should some of this go into the parent Optimizer class, such as the boundary acquiring?
+    modded = False
+    for var in self.toBeSampled:
+      dist = self.distDict[var]
```

Comment:
We will have to add this back if we want to do statistical robust optimization; since this behaves exactly like the Sampler does, and the Optimizer is a Sampler, and it would take a seperate part of code just to deal with "do I care about distributions or not", I recommend changing this to be a future development and not delay this PR.